src/write.c

```diff
@@ -1219,10 +1224,173 @@ static avifResult avifEncoderAddImageInternal(avifEncoder * encoder,
+    }
+
+    // avifCodecEncodeImageFunc() uses a quantizer argument instead of the avifEncoder.quality one.
+    // The quantizer range [0:63] being smaller than the quality range [0:100], it would be faster
+    // to operate the binary search below directly on the quantizer setting rather than on the quality one.
+    // For simplicity, it uses the quality for now.
+
+    // Use the encoder instance as a "current settings" holder.
+    const avifBool searchForQuality = encoder->quality == AVIF_QUALITY_DEFAULT;
+    const avifBool searchForQualityAlpha = encoder->qualityAlpha == AVIF_QUALITY_DEFAULT;
+
+    int closestQuality = AVIF_QUALITY_DEFAULT;
+    avifRWData closestEncoded = AVIF_DATA_EMPTY;
+    size_t closestSizeDiff = 0;
+    avifIOStats closestIoStats = { 0, 0 };
+
+    int minQuality = AVIF_QUALITY_WORST; // inclusive
+    int maxQuality = AVIF_QUALITY_BEST;  // inclusive
+    while (minQuality <= maxQuality) {
+        const int quality = (minQuality + maxQuality) / 2;
+        if (searchForQuality) {
+            encoder->quality = quality;
+        }
+        if (searchForQualityAlpha) {
+            encoder->qualityAlpha = quality;
+        }
+
+        avifRWData encoded = AVIF_DATA_EMPTY;
+        avifIOStats ioStats = { 0, 0 };
+        const avifResult result = avifEncoderWriteInternal(encoder,
+                                                           gridCols,
+                                                           gridRows,
+                                                           cellImages,
+                                                           durationInTimescales,
+                                                           addImageFlags,
+                                                           &encoded,
+                                                           &ioStats,
+                                                           &encoder->diag);
+        if (result != AVIF_RESULT_OK) {
+            avifRWDataFree(&encoded);
+            avifRWDataFree(&closestEncoded);
+            return result;
+        }
+
+        size_t sizeDiff;
+        if (encoded.size > encoder->targetSize) {
+            sizeDiff = encoded.size - encoder->targetSize;
+            maxQuality = quality - 1;
+        } else {
+            sizeDiff = encoder->targetSize - encoded.size;
+            minQuality = quality + 1;
+        }
+
+        if ((closestQuality == AVIF_QUALITY_DEFAULT) || (sizeDiff < closestSizeDiff)) {
+            closestQuality = quality;
+            avifRWDataFree(&closestEncoded);
+            closestEncoded = encoded;
+            encoded.size = 0;
+            encoded.data = NULL;
+            closestSizeDiff = sizeDiff;
+            closestIoStats = ioStats;
+        } else {
+            avifRWDataFree(&encoded);
+        }
+
+        if (sizeDiff == 0) {
+            break;
+        }
+    }
+
+    // Output the used quality directly in the avifEncoder struct.
```

Comment:
I refrained from doing this with the `autoTiling`  feature. That was a somewhat arbitrary decision, to avoid modifying the `avifEncoder` struct that we received from the user. We should modify the `autoTiling` code to also output the used `tileRowsLog2` and `tileColsLogs` directly in the `avifEncoder` struct, and print the used values in avifenc.

Reply:
As noted in https://github.com/AOMediaCodec/libavif/pull/1278#issue-1573254420, I am not sure about using the targetSize API in avifenc.c. I think displaying what is going on to stdout is important, and thus I would lean towards using both #1278 and #1274 implementations.

With that said, maybe there is no need to modify `avifEncoder.quality`. An alternative is outputting the final `quality` and `qualityAlpha` through `avifIOStats`. I did not do that at first because it gets confusing when quantizers are given, but it should be possible.

What do you think?

Reply:
Outputting the achieved quality might be good. As for progress, another option would be to introduce a callback like webp to indicate progress.

Reply:
> Outputting the achieved quality might be good.

I assume you meant through `avifIOStats`. See #1286.

> As for progress, another option would be to introduce a callback like webp to indicate progress.

This is significantly more work than planned. But it may be necessary to be able to cancel the encoding, for native AVIF support in Photoshop for example.